src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1003,6 +1003,154 @@ void Lowering::LowerFusedMultiplyAdd(GenTreeHWIntrinsic* node)
     }
 }
 
+//----------------------------------------------------------------------------------------------
+// TryLowerConstVec: Unfold the Vector to Broadcast->CreateScalarUnsafe->ConstScalar form
+//                   when the constant vector has all the same elements and is an operand
+//                   of an embedded broadcast compatible instruction.
+//
+//  Arguments:
+//     node - The constant vector node
+GenTree* Lowering::TryLowerConstVec(GenTreeVecCon* node)
+{
+    LIR::Use            use;
+    bool                foundUse         = BlockRange().TryGetUse(node, &use);
+    GenTreeHWIntrinsic* embBroadcastNode = nullptr;
+    if (foundUse && use.User()->OperIs(GT_HWINTRINSIC) && use.User()->AsHWIntrinsic()->OperIsEmbBroadcastCompatible())
+    {
+        embBroadcastNode                = use.User()->AsHWIntrinsic();
+        var_types   simdType            = embBroadcastNode->TypeGet();
+        var_types   simdBaseType        = embBroadcastNode->GetSimdBaseType();
+        CorInfoType simdBaseJitType     = embBroadcastNode->GetSimdBaseJitType();
+        bool        isCreatedFromScalar = true;
+
+        int ElementCount = GenTreeVecCon::ElementCount(genTypeSize(simdType), simdBaseType);
+        switch (simdBaseType)
+        {
+            case TYP_FLOAT:
+            case TYP_INT:
+            case TYP_UINT:
+            {
+                uint32_t FirstElement = static_cast<uint32_t>(node->gtSimdVal.u32[0]);
+                for (int i = 1; i < ElementCount; i++)
+                {
+                    uint32_t ElementToCheck = static_cast<uint32_t>(node->gtSimdVal.u32[i]);
+                    if (FirstElement != ElementToCheck)
+                    {
+                        isCreatedFromScalar = false;
+                        break;
+                    }
+                }
+                // There are special case when all elements in the vector are 1/0,
+                // there are exsiting optimization for those case, filter them out
+                // of the embedded broadcast.
+                if (isCreatedFromScalar && (FirstElement == UINT32_MAX || FirstElement == UINT32_MIN))
+                {
+                    isCreatedFromScalar = false;
+                }
```

Comment:
Seems like this could be an early exit instead, avoiding the loop entirely for `zero` and `allbitsset` cases